src/ContainerTrait.php

```diff
@@ -14,22 +39,152 @@ public function _unique_element($desired = null)
         return $desired.($postfix > 1 ? ('_'.$postfix) : '');
     }
 
-    protected function _unique() {
+    /**
+     * Extension to add() method which will perform linking of
+     * the object with the current class.
+     */
+    protected function _add_Container($element, $args = [])
+    {
+        // Normalize the arguments, bring name out
+        if (is_string($args)) {
+
+            // passed as string
+            $args=[$args];
+        } elseif (!is_array($args)) {
+
+            throw new Exception(['Second argument must be array','arg2'=>$args]);
+        } elseif (is_array($args) && isset($args['name'])) {
+
+            // passed as ['name'=>'foo'];
+            $args[0]=$args['name'];
+            unset($args['name']);
+        } elseif (isset($element->short_name)) {
+
+            // element has a name already
+            $args[0]=$element->short_name;
+        } elseif (isset($element->_trackableTrait)) {
+
+            // ask element on his preferred name, then make it unique.
+            $cn = $element->getDesiredName();
+            $args[0] = $this->_unique_element($cn);
+        } else {
+
+            // generate name based on the class
+            $cn = str_replace('\\', '_', strtolower(get_class($element)));
+            $args[0] = $this->_unique_element($cn);
+        }
+
```

Comment:
It looks that we have to use $this->_unique_element($desired_name) always not only in some particular cases.
That's because I can set equal names for 2 objects and then they will collide in $elements array if not making call to _unique_element().